File: rust/crates/corlinman-channels/src/rate_limit.rs
//! Per-key token-bucket rate limiter used by [`crate::router::ChannelRouter`].
//!
//! The router consults one bucket per rate-limit dimension (per-group,
//! per-sender). The key is a stable string derived from the channel binding
//! (`"qq:group:<gid>"`, `"qq:sender:<gid>:<uid>"`) so collisions across
//! channels / threads are impossible.
//!
//! Algorithm: classic token bucket with linear refill.
//! - `capacity = per_min` (so a freshly-seen key can burst up to `per_min`
//!   turns instantly).
//! - `refill_per_sec = per_min / 60` (linear, not jittered).
//! - `check`: refill based on wall clock, try to consume 1 token, return
//!   `true` iff the bucket had ≥ 1 token.
//!
//! This is a per-process limiter. See the module-level TODOs for the Redis
//! variant we want once a second gateway replica ships.
//!
//! GC: the internal [`DashMap`] grows unboundedly if we never prune. The
//! [`TokenBucket::start_gc`] task periodically sweeps entries whose
//! `last_refill` is more than an hour old — idle groups / senders drop out
//! and re-appear on the next message at full capacity (semantically fine: a
//! group that hasn't spoken in an hour is not mid-burst anyway).

use std::sync::Arc;
use std::time::{Duration, Instant};

use dashmap::DashMap;
use tokio::task::JoinHandle;
use tokio_util::sync::CancellationToken;

/// Entries older than this are pruned by the background GC sweeper.
const GC_STALE_AFTER: Duration = Duration::from_secs(3600);

/// How often the background sweeper walks the map.
const GC_INTERVAL: Duration = Duration::from_secs(300);

/// A thread-safe, per-key token bucket.
pub struct TokenBucket {
    /// Max tokens per key (equals `per_min`).
    capacity: f64,
    /// Tokens replenished per second (equals `per_min / 60.0`).
    refill_per_sec: f64,
    state: DashMap<String, BucketState>,
}

struct BucketState {
    tokens: f64,
    last_refill: Instant,
}

impl TokenBucket {
    /// Build a bucket that allows `per_min` events per minute per key.
    pub fn per_minute(per_min: u32) -> Self {
        let capacity = per_min as f64;
        let refill_per_sec = capacity / 60.0;
        Self {
            capacity,
            refill_per_sec,
            state: DashMap::new(),
        }
    }

    /// Try to consume 1 token from `key`'s bucket. Returns `true` if the
    /// caller is allowed to proceed. A brand-new key starts at full capacity.
    pub fn check(&self, key: &str) -> bool {
        let now = Instant::now();
        let mut entry = self.state.entry(key.to_string()).or_insert(BucketState {
            tokens: self.capacity,
            last_refill: now,
        });
        let elapsed = now.duration_since(entry.last_refill).as_secs_f64();
        entry.tokens = (entry.tokens + elapsed * self.refill_per_sec).min(self.capacity);
        entry.last_refill = now;
        if entry.tokens >= 1.0 {
            entry.tokens -= 1.0;
            true
        } else {
            false
        }
    }

    /// Number of live keys currently tracked — useful for tests and future
    /// metrics.
    pub fn tracked_keys(&self) -> usize {
        self.state.len()
    }

    /// Spawn a background task that periodically sweeps stale keys
    /// (`last_refill > GC_STALE_AFTER` ago). Returns the join handle so
    /// callers can abort on shutdown. Cancellation via `cancel` stops the
    /// loop cleanly.
    pub fn start_gc(self: Arc<Self>, cancel: CancellationToken) -> JoinHandle<()> {
        tokio::spawn(async move {
            let mut ticker = tokio::time::interval(GC_INTERVAL);
            // Skip the immediate tick — nothing to sweep yet.
            ticker.tick().await;
            loop {
                tokio::select! {
                    biased;
                    _ = cancel.cancelled() => break,
                    _ = ticker.tick() => {
                        self.sweep_stale();
                    }
                }
            }
        })
    }

    /// Remove entries whose `last_refill` is older than [`GC_STALE_AFTER`].
    /// Exposed for tests; the background sweeper calls this on each tick.
    pub fn sweep_stale(&self) {
        let cutoff = Instant::now() - GC_STALE_AFTER;
        self.state.retain(|_, v| v.last_refill >= cutoff);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bucket_allows_within_capacity() {
        let b = TokenBucket::per_minute(20);
        for _ in 0..20 {
            assert!(b.check("g:1"));
        }
    }

    #[test]
    fn bucket_denies_when_empty() {
        let b = TokenBucket::per_minute(20);
        for _ in 0..20 {
            assert!(b.check("g:1"));
        }
        // 21st immediately after exhausting → refill ≪ 1 token → deny.
        assert!(!b.check("g:1"));
    }

    #[test]
    fn bucket_refills_over_time() {
        // 60/min = 1 token per second. Exhaust then wait > 1s for a refill.
        let b = TokenBucket::per_minute(60);
        for _ in 0..60 {
            assert!(b.check("g:1"));
        }
        assert!(!b.check("g:1"));
        std::thread::sleep(Duration::from_millis(1100));
        assert!(b.check("g:1"), "expected refill after 1.1s at 1 token/sec");
    }

    #[test]
    fn bucket_isolates_keys() {
        let b = TokenBucket::per_minute(3);
        for _ in 0..3 {
            assert!(b.check("a"));
        }
        assert!(!b.check("a"));
        // Key "b" has its own bucket, unaffected.
        for _ in 0..3 {
            assert!(b.check("b"));
        }
        assert!(!b.check("b"));
    }

    #[test]
    fn bucket_capacity_caps_refill() {
        // A bucket idle for a long time should not accumulate beyond capacity.
        let b = TokenBucket::per_minute(5);
        // Force-seed an idle key by checking once to create state, then
        // rewind last_refill far enough to accumulate >> capacity worth.
        assert!(b.check("k"));
        {
            let mut s = b.state.get_mut("k").unwrap();
            s.last_refill = Instant::now() - Duration::from_secs(3600);
            s.tokens = 0.0;
        }
        // One refill computation on next check → should cap at capacity,
        // not at refill_per_sec * elapsed = 300.
        assert!(b.check("k"));
        // After consuming 1, the bucket should have exactly capacity-1.
        let remaining = b.state.get("k").unwrap().tokens;
        assert!(
            (remaining - (b.capacity - 1.0)).abs() < 1e-6,
            "expected refill capped at capacity, got {remaining}"
        );
    }

    #[test]
    fn sweep_drops_stale_keys() {
        let b = TokenBucket::per_minute(5);
        assert!(b.check("idle"));
        assert_eq!(b.tracked_keys(), 1);
        // Rewind last_refill past the stale cutoff.
        {
            let mut s = b.state.get_mut("idle").unwrap();
            s.last_refill = Instant::now() - GC_STALE_AFTER - Duration::from_secs(1);
        }
        b.sweep_stale();
        assert_eq!(b.tracked_keys(), 0);
    }
}

File: python/packages/corlinman-channels/src/corlinman_channels/rate_limit.py
"""Per-key token-bucket rate limiter used by :mod:`corlinman_channels.router`.

Python port of ``rust/.../rate_limit.rs``. The router consults one bucket
per rate-limit dimension (per-group, per-sender). The key is a stable
string derived from the channel binding (``"qq:group:<gid>"``,
``"qq:sender:<gid>:<uid>"``) so collisions across channels / threads are
impossible.

Algorithm: classic token bucket with linear refill.

- ``capacity = per_min`` (a freshly-seen key can burst up to ``per_min``
  turns instantly).
- ``refill_per_sec = per_min / 60`` (linear, not jittered).
- :meth:`TokenBucket.check`: refill based on monotonic clock, try to
  consume 1 token, return ``True`` iff the bucket had >= 1 token.

This is a per-process limiter. See the module-level TODOs in the Rust
crate for the Redis variant once a second gateway replica ships.

GC: the internal map grows unboundedly if we never prune.
:meth:`TokenBucket.start_gc` schedules a background sweeper task that
prunes entries whose ``last_refill`` is more than an hour old — idle
groups / senders drop out and re-appear on the next message at full
capacity (semantically fine: a group that hasn't spoken in an hour is
not mid-burst anyway).

## Deliberate deviations from Rust

- Rust uses ``dashmap::DashMap`` for lock-free shared access. Python's
  asyncio is single-threaded so a plain ``dict`` plus an ``asyncio.Lock``
  around mutate paths is equivalent. We use ``time.monotonic()`` instead
  of ``Instant``; both are wall-clock-immune.
- Rust returns a ``tokio::task::JoinHandle``; we return ``asyncio.Task``
  (``start_gc`` is now an ``async`` method because creating tasks
  requires an event loop).
- Rust's ``CancellationToken`` becomes an ``asyncio.Event`` (``cancel.set()``
  takes the place of ``cancel.cancel()``); callers either set it
  directly or rely on task cancellation.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass

__all__ = [
    "GC_INTERVAL",
    "GC_STALE_AFTER",
    "TokenBucket",
]


#: Entries older than this are pruned by the background GC sweeper.
#: Matches ``GC_STALE_AFTER`` in Rust ``rate_limit.rs`` (1 hour).
GC_STALE_AFTER: float = 3600.0

#: How often the background sweeper walks the map. Matches ``GC_INTERVAL``
#: in Rust ``rate_limit.rs`` (5 minutes).
GC_INTERVAL: float = 300.0


@dataclass(slots=True)
class _BucketState:
    """Per-key token-bucket state. Mirrors ``BucketState`` in Rust."""

    tokens: float
    last_refill: float


class TokenBucket:
    """Thread-safe per-key token bucket.

    Mirrors the Rust ``TokenBucket`` struct field-for-field with the
    deviations documented at the module level. The public surface
    (:meth:`check`, :meth:`tracked_keys`, :meth:`sweep_stale`,
    :meth:`start_gc`) is intentionally identical.

    Construct via :meth:`per_minute` rather than the bare constructor so
    the ``refill_per_sec`` arithmetic matches Rust exactly:

    >>> b = TokenBucket.per_minute(60)
    >>> b.check("group:1")
    True
    """

    __slots__ = ("_capacity", "_refill_per_sec", "_state")

    def __init__(self, capacity: float, refill_per_sec: float) -> None:
        """Low-level constructor. Prefer :meth:`per_minute`.

        Exposed primarily so future variants (per-hour, per-day) can be
        constructed without going through the ``per_minute`` shortcut.
        """
        self._capacity: float = capacity
        self._refill_per_sec: float = refill_per_sec
        self._state: dict[str, _BucketState] = {}

    @classmethod
    def per_minute(cls, per_min: int) -> TokenBucket:
        """Build a bucket that allows ``per_min`` events per minute per key."""
        capacity = float(per_min)
        return cls(capacity=capacity, refill_per_sec=capacity / 60.0)

    @property
    def capacity(self) -> float:
        """Per-key capacity (= ``per_min`` for buckets built via
        :meth:`per_minute`). Exposed mainly for testing parity with the
        Rust ``self.capacity`` field access."""
        return self._capacity

    def check(self, key: str) -> bool:
        """Try to consume 1 token from ``key``'s bucket.

        Returns ``True`` if the caller is allowed to proceed. A brand-new
        key starts at full capacity. Mirrors ``TokenBucket::check`` in
        Rust.
        """
        now = time.monotonic()
        entry = self._state.get(key)
        if entry is None:
            entry = _BucketState(tokens=self._capacity, last_refill=now)
            self._state[key] = entry
        elapsed = now - entry.last_refill
        if elapsed < 0:
            # Monotonic clock should not regress, but be defensive — clamp
            # to 0 so a single weird tick doesn't leak negative tokens.
            elapsed = 0.0
        entry.tokens = min(
            entry.tokens + elapsed * self._refill_per_sec,
            self._capacity,
        )
        entry.last_refill = now
        if entry.tokens >= 1.0:
            entry.tokens -= 1.0
            return True
        return False

    def tracked_keys(self) -> int:
        """Number of live keys currently tracked.

        Useful for tests and future metrics. Matches
        ``TokenBucket::tracked_keys`` in Rust.
        """
        return len(self._state)

    def sweep_stale(self) -> None:
        """Remove entries whose ``last_refill`` is older than
        :data:`GC_STALE_AFTER`. Exposed for tests; the background sweeper
        calls this on each tick. Matches ``TokenBucket::sweep_stale``.
        """
        cutoff = time.monotonic() - GC_STALE_AFTER
        # Materialize the key list first so we don't mutate during iteration.
        stale = [k for k, v in self._state.items() if v.last_refill < cutoff]
        for k in stale:
            self._state.pop(k, None)

    def start_gc(
        self,
        cancel: asyncio.Event | None = None,
        interval: float = GC_INTERVAL,
    ) -> asyncio.Task[None]:
        """Spawn a background task that periodically calls
        :meth:`sweep_stale`.

        Mirrors ``TokenBucket::start_gc`` in Rust. Pass an
        :class:`asyncio.Event` and ``set()`` it to stop the loop cleanly;
        otherwise rely on :meth:`asyncio.Task.cancel`. The ``interval``
        argument lets tests use a sub-second cadence without modifying
        the module-level default.

        Unlike the Rust version, this method takes the event loop as
        implicit context (``asyncio.create_task`` requires a running
        loop). Call it from inside an async context (e.g. an adapter's
        ``connect()``).
        """
        cancel_event = cancel if cancel is not None else asyncio.Event()

        async def _loop() -> None:
            try:
                # Skip the immediate tick — nothing to sweep yet (matches
                # Rust ``ticker.tick().await`` before the loop body).
                await self._wait_for_tick(cancel_event, interval)
                while not cancel_event.is_set():
                    self.sweep_stale()
                    await self._wait_for_tick(cancel_event, interval)
            except asyncio.CancelledError:
                # Standard shutdown path — exit silently.
                return

        return asyncio.create_task(_loop(), name="token-bucket-gc")

    @staticmethod
    async def _wait_for_tick(cancel: asyncio.Event, interval: float) -> None:
        """Wait ``interval`` seconds or until ``cancel`` is set.

        Equivalent of Rust's ``tokio::select! { _ = cancel.cancelled() =>
        break, _ = ticker.tick() => ... }`` — whichever fires first wins
        and the caller observes via ``cancel.is_set()``.
        """
        try:
            await asyncio.wait_for(cancel.wait(), timeout=interval)
        except TimeoutError:
            return
